fix(ast): estree compat `CatchClause`

hi-ogawa opened the pull request:
Part of https://github.com/oxc-project/oxc/issues/2854

```diff
-Positive Passed: 28924/44293 (65.30%)
+Positive Passed: 30446/44293 (68.74%)
```

hi-ogawa commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8975?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8975** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8975?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8975?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8974** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8974?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8973** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8973?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#8972** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8972?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8975).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hi-ogawa commented on tasks/ast_tools/src/generators/typescript.rs:
```diff
@@ -224,7 +224,8 @@ fn generate_ts_type_def_for_struct_field<'s>(
         }
 
-        if field_type_name.contains('|') {
+        // need `type` instead of `interface` when flattening BindingPattern
+        if field_type_name.contains('|') || field_type_name == "BindingPattern" {
```
Hard-coding `BindingPattern` is bad, but I didn't find a good way to force `output_as_type`. I can look into supporting this case in general later.

overlookmotel replied:
Yes, not ideal in long term, but totally fine for now. Personally, I think we should look at changing the Rust types for `BindingPattern` anyway. It's a bit weird compared to all the other AST types - it's the only node which doesn't have it's own `Span` but gets its span from a child, for example.

overlookmotel commented:
### Merge activity

* **Feb 9, 1:59 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
From 14% to 68% passing in 4 short PRs!

hi-ogawa committed:
fix(ast): estree compat `CatchClause` (#8975)

Part of https://github.com/oxc-project/oxc/issues/2854

```diff
-Positive Passed: 28924/44293 (65.30%)
+Positive Passed: 30446/44293 (68.74%)
```
